runtime/lua/vim/lsp/codelens.lua

```diff
@@ -0,0 +1,223 @@
+M.__namespaces = namespaces
+
+
+--- Return all lenses for the given buffer
+---@return table (`CodeLens[]`)
+function M.get(bufnr)
+  local lenses_by_client = lens_cache_by_buf[bufnr]
+  if not lenses_by_client then return {} end
+  local lenses = {}
+  for _, client_lenses in pairs(lenses_by_client) do
+    vim.list_extend(lenses, client_lenses)
+  end
+  return lenses
+end
+
+
+--@private
+local function execute_lens(lens, bufnr, client_id)
+  local line = lens.range.start.line
+  api.nvim_buf_clear_namespace(bufnr, namespaces[client_id], line, line + 1)
+
+  -- Need to use the client that returned the lens → must not use buf_request
+  local client = vim.lsp.get_client_by_id(client_id)
+  assert(client, 'Client is required to execute lens, client_id=' .. client_id)
+  client.request('workspace/executeCommand', lens.command, function(...)
+    local result = vim.lsp.handlers['workspace/executeCommand'](...)
+    M.refresh()
+    return result
+  end, bufnr)
+end
+
+
+--- Run the code lens in the current line
+---
+function M.run()
+  local line = api.nvim_win_get_cursor(0)[1]
+  local bufnr = api.nvim_get_current_buf()
+  local options = {}
+  local lenses_by_client = lens_cache_by_buf[bufnr] or {}
+  for client, lenses in pairs(lenses_by_client) do
+    for _, lens in pairs(lenses) do
+      if lens.range.start.line == (line - 1) then
+        table.insert(options, {client=client, lens=lens})
+      end
+    end
+  end
+  if #options == 0 then
+    print('No executable codelens found at current line')
+  elseif #options == 1 then
+    local option = options[1]
+    execute_lens(option.lens, bufnr, option.client)
+  else
+    local options_strings = {"Code lenses:"}
+    for i, option in ipairs(options) do
+       table.insert(options_strings, string.format('%d. %s', i, option.lens.command.title))
+    end
+    local choice = vim.fn.inputlist(options_strings)
+    if choice < 1 or choice > #options then
+      return
+    end
+    local option = options[choice]
+    execute_lens(option.lens, bufnr, option.client)
+  end
+end
+
+
+--- Display the lenses using virtual text
+---
+---@param lenses table of lenses to display (`CodeLens[] | null`)
+---@param bufnr number
+---@param client_id number
+function M.display(lenses, bufnr, client_id)
+  if not lenses or not next(lenses) then
+    return
+  end
+  local lenses_by_lnum = {}
+  for _, lens in pairs(lenses) do
+    local line_lenses = lenses_by_lnum[lens.range.start.line]
+    if not line_lenses then
+      line_lenses = {}
+      lenses_by_lnum[lens.range.start.line] = line_lenses
+    end
+    table.insert(line_lenses, lens)
+  end
+  local ns = namespaces[client_id]
+  local num_lines = api.nvim_buf_line_count(bufnr)
+  for i = 0, num_lines do
+    local line_lenses = lenses_by_lnum[i]
+    api.nvim_buf_clear_namespace(bufnr, ns, i, i + 1)
+    local chunks = {}
+    for _, lens in pairs(line_lenses or {}) do
+      local text = lens.command and lens.command.title or 'Unresolved lens ...'
```

Comment:
Maybe a future improvement, but what do you think about adding a way to add in a prefix symbol similar to the way diagnostics is done. I know for myself I'd love to easily be able to have a `▸` preface or something denoting a lens.

Reply:
Sounds good as a follow up. Maybe utilizing the `vim.lsp.config` from another pending PR, other via the handler `lsp.with` configuration.